File: LeetCode/206. 反转链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  let p = head,
    pre = null;
  while (p) {
    let next = p.next;
    p.next = pre;
    pre = p;
    p = next;
  }
  return pre;
};

File: LeetCode/206. 反转链表.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
  pub fn reverse_list(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
      let mut pre=None;
      while head.is_some(){
          let mut n=head.unwrap();
          head=n.next;
          n.next=pre;
          pre=Some(n);
      }
      return pre;
  }
  
    pub fn reverse_list1(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut oldHead = head;
        let mut newHead = None;
        while let Some(mut node) = oldHead {
            oldHead = node.next.take();
            node.next = newHead;
            newHead = Some(node);
        }
        newHead
    }
}

